### PR TITLE
PriorityScheduler: reduce unpark calls

### PR DESCRIPTION
**Pull request by jentfoo**
A best effort to reduce unpark calls seems to provide significant performance boost in the benchmarks.

We flip a boolean when we are about to unpark the head thread.  This prevents other threads from attempting to unpark any other (but likely the same) threads until one wakes up.  As soon as a (any) thread wakes up it flips the boolean back so that the next parked thread can be woken up if needed.

One key detail in how this works safely is that the boolean must be set back to false BEFORE the loop continues and it examines for the next item.  If we tried to expand this unpark blackout window to while we were checking for tasks we would have a potential race condition that could cause a thread to sleep forever.

I also could not find a way to safely correlate the unsetting of the blackout to what thread actually wakes up.  In theory if we could continue the unpark blackout until after the thread unparked actually wakes up (rather than any thread) there may be some gains to be had.  However this really complicates things significantly, and I could not at this time find a way to do it safely.

In part this makes a lot of sense for threadly because we will only unpark the head of the parked queue, and thus we likely were unparking the same thread over and over again.

### CODE REVIEW

**Comment by jentfoo** [comment]
So I have done a lot of benchmarking on this over the night.  The results are quite interesting:

As of the current state of this PR ( 4e37f40 ) we have the most performance gains in litesockets based benchmarks.  We have almost no gains in the threadly benchmarks (but there is some).

It seems that the threadly benchmarks were getting the biggest bump by reducing the concurrent workers in `workersIdle`.  When we switched from a global unpark lock to a per-worker unpark lock the gains in the threadly benchmarks almost completely evaporated.

However litesockets performed very very poorly in commit b3204f6.  It seems that it really wants more concurrency in `workersIdle` so that presumably execute tasks can be consumed quicker.

This also explains why the gains were so large for the scheduled task performance.  Because scheduled tasks require locking (which execute tasks do not), reducing the `workersIdle` concurrency allowed for less lock contention.

This actually points to the fact that in real world applications our scheduled task performance is probably better than our benchmarks indicate, when you consider it being mixed in with primarily execute tasks.

I have tried to play with hybrid algorithms, but so far have not been able to find any way to get the best of both worlds.

TL;DR: Because I trust the litesockets load patterns more as a real world pattern, and because the changes as currently in this PR show no decreases anywhere, I think we should go forward with what we have now.

**Comment by jentfoo** [comment]
Per my last comment, merging, but feel free to comment if additional ideas or considerations come up.

**Comment by chrisvest** [comment]
If you're on Linux, park & unpark moves processes between state queues (runnable, waiting, etc.) in the kernel. Modifications to these queues are guarded by spin-locks. Thus it makes sense that multi-threaded programs run faster when they avoid unnecessary calls to park/unpark, in case you were wondering. You'll be able to see this when profiling with `perf` and have kernel symbols enabled.
